ucan-dns+web.md

```diff
@@ -0,0 +1,209 @@
+
+# Abstract
+
+The [UCAN specification][ucan] describes token principals via [`did:key`][] method and allows other DID methods to be used. This specification defines extension for use of [`did:dns`][] and [`did:web`] methods.
+
+## Motivation
+
+While [`did:key`][] method is excellent for cryptographic verifiability, it does however suffer in contexts where key rotation is desired in way that would not invalidating all delegatinon chains it is part of. Following scenario describes these limitions:
+
+Alice with DID `did:key:zAlice` wants to delegate some capability to a `ucan.store` service, so it can redelegate that capability to another agent `did:key:zAli`. If `ucan.store` service rotates a key it will no longer be able to issue redelagation.
+
+<!--
+ with DID `did:key:zPreRotation`. User Aice requests access to delegated capability on another device with DID `did:key:zAli` from the `ucan.store` _(after going through an out of bound verification)_. Unfortunatelly `ucan.store` had to rotate key (because old on was compromised) and now it has new DID `did:key:zPostRotation`. It is impossible for `ucan.store` to delegate capbility back to the user. Delegation can only be arranged if user can issue new delegation from `did:key:zAlice`, but user no longer has access to the that key delegation is no longer possible.
+
+> There are other ways Alice could arrange delegation from `did:key:zAlice` to `did:key:zAli` with different tradeoffs. Evaluating them would be a good idea for a system designer, here we defining solution when no other option offers desired tradeoffs. -->
+
+To address described limitation we propose use of [`did:dns`] or [`did:web`] principals so that delegation from `did:key:zAlice` no loner is tied to a specific key. This would get us step further, but still we run into a problem, after key rotation delegation to `did:key:zAli` is no longer valid as the key that signed it is no longer the one did document resolves to.
+
+This specification describes solution to the second order problem by requiring that [`did:key`] that [`did:dns`] and [`did:web`] resolve to MUST assume an ambient authority over (pre-resolution) DID, which it MAY delegate to other principals through standard UCAN delegation.
+
+> Expample below illustrates `did:dns:w3.storage` delegating own resource to `did:key:zService`, which in turn redelegates it to `did:key:zRotation`.
+>
+> This setup allows primary key (one that DID document resolves to) to be kept very safe e.g. on a piece of paper in safe deposit box, is it is only needed to delegate capabilities to `did:key:zService`. That key also is only used for rating keys and therefor can be stored securily e.g. in hardware key.
+
+```ts
+{
+  aud: "did:key:zRotation",
+  iss: "did:key:zService",
+  exp: null,
+  att: [],
+  prf: [{
+    iss: "did:dns:w3.storage",
+    aud: "did:key:zService",
+    exp: null
+    att: []
+  }]
+}
+```
```

Comment:
Because `att: []` means "delegate nothing", I think this doesn't work. Are you sure that you don't want delegation?


```json
{
  iss: "did:key:zService",
  aud: "did:key:zRotation",
  exp: null,
  att: [{
    iss: "did:dns:w3.storage",
    aud: "did:key:zService",
    exp: null
    att: []
  }],
  prf: [ fromOldKeyToService ]
}
```

Reply:
@matheus23 also came up with something for WNFS recently that may be interesting here (adapted to DID management)

```js
{
  iss: "did:dns:example",
  aud: "did:key:zService",
  exp: null,
  att: [{
    with: "did:dns:example",
    can: "did/rotate",
    nb: { to: "did:key:zNewKey" }
  }]
}
```

Here the UCAN only allows the DID document to be updated in a very specific way.

Per my earlier comment about revocation above, this would still require checking the list of valid public keys at read time. I don't think that you can get around that if you support revocation though.

Reply:
> Are you sure that you don't want delegation?

Another thing worth mentioning about the reasoning here: if the service can go rogue and change the key to anyone (including itself!), it's already equivalent to it having `*` rights to the delegated account. I don't think that you gain anything by making it a proof instead of a delegation.

Reply:
> Because att: [] means "delegate nothing", I think this doesn't work. Are you sure that you don't want delegation?

I was kind of overloading it to mean redelegate everything it may be delegated to in the future.

> @matheus23 also came up with something for WNFS recently that may be interesting here (adapted to DID management)

How interesting, in some iteration of this document I had the same thing (although I called it `did/set` instead). However I have went back on that idea, because it felt too domain specific, I felt that it was bit confusing that ability to update DID key implied that I could delegate any capability delegated to that did.

I mean yes, but I wish UCAN had a more explicit thing than relying on DID specifics

> Note that in this case the DID itself is in the with field because you're updating the DID itself!

I have also considered following 

```js
{
  aud: "did:key:zRotation",
  iss: "did:key:zService",
  exp: null,
  att: [{ with: "did:dns:example", can: "*" }],
  prf: [{
    iss: "did:dns:w3.storage",
    aud: "did:key:zService",
    exp: null
    att: [{ with: "did:dns:example", can: "*" }],
  }]
}
```

But then concluded it was not right, because `{ with: "did:dns:example", can: "*" }` does not exactly imply `{ with: "did:key:zAlice", can: "*" } ` even if later was delegated to `"did:dns:example"`, or maybe it does but somehow it does not feel very clear to me.


I think this would be the clearest and most explicit that reads as `did:key:zService` can do anything with everything that `did:dns:w3.storage`. Just like `file/*` implies all current and future abilities under `file/` namespace so `with: "*"` could imply all the current and future resources principal has authority over.


```js
{
  iss: "did:dns:w3.storage",
  aud: "did:key:zService",
  exp: null,
  att: [{ with: "*", can: "*" }],
}
```

Yet another alternative would be to make `att` optional to imply an unattenuated delegation

```js
{
  iss: "did:dns:w3.storage",
  aud: "did:key:zService",
  exp: null,
}
```

Reply:
Oh, does this mean exactly what I was trying to express with various other forms ?

```js
{
  iss: "did:key:zService",
  aud: "did:key:zRotation",
  exp: null,
  att: [{
    with: "own://did:key:zOldKey/*",
    can: "*"
  }],
  prf: [ fromOldKeyToService ]
}
```

Or in other words does `did:key:zAlice` would be included there if delegation with `did:key:zAlice` to `did:key:zOldKey` exists ?

Reply:
> I mean yes, but I wish UCAN had a more explicit thing than relying on DID specifics

Tell me more! Can you describe more about the kind of properties that you wish it had?

Reply:
I pull relevant discussion into https://github.com/ucan-wg/spec/issues/131, probably better to continue this discussion there.